Caching split structure docs in memcached

ormsbee opened the pull request:
@cpennington, @doctoryes: Another possibility for reducing Mongo load. Please consider for load testing.

cpennington commented:
`EdxJSONEncoder` will do `UsageKeys` and `CourseKeys` (although I'm surprised you needed to. They're just stored as strings in the mongo document).

cpennington commented:
How were you measuring the timing?

ormsbee commented:
I was using pyinstruments in setprofile mode.

ormsbee commented:
Switched to zlib and the the performance seems to be in line with what we already do. This would reduce load on Mongo in exchange for heavier usage of memcached and increased CPU utilization the web workers. Still have to test more to be confident that it hasn't slowed things down.

ormsbee commented:
So while the profiling results give similar numbers, response times are slightly higher with the profiling turned off (~10% slower). I'm going to experiment a little more after my next meeting.

cpennington commented:
@ormsbee we noticed while doing more digging that

ormsbee commented:
Lowering the compression level helped somewhat. I'm going to fix up the tests -- I think they're mostly failing because the query count went down in a number of places.

ormsbee commented:
Some rough index page timings on my local machine:

```
this branch:

# MRev
2.796
2.776
2.884
2.868
2.804

# Demo Course
0.504
0.520
0.672
0.528
0.640

master:

# MRev
2.832
2.724
2.808
2.636
2.632

# Demo Course
0.664
0.764
0.612
0.632
0.728
```

So somewhat better for small courses and somewhat worse for large courses -- but my understanding is that we have a lot of headroom on memcached. I'd be curious how this plays out on EC2 where I expect more CPU and higher latency costs.

cpennington commented:
Yeah, we should run this through the same kind of loadtest that we ran https://github.com/edx/edx-platform/pull/8276 through.

ormsbee commented:
@cpennington: I've cleaned this up and rebased. Where would you like to go with this PR? I'm happy to address comments/issues, but I'm probably not going to have the time to run it through the load test environment.

ormsbee commented:
test this please

cpennington commented:
I think these tests may end up vulnerable to cache eviction (since they're counting queries, but those counts are dependent on whether the structure is cached or not).

I think we should wait to load test this until we have a local-to-the-webhead memcache instance (@e0d).

ormsbee commented:
The entries never expire, so if it's using an in-memory cache for tests (it is, right?), we shouldn't have to worry about eviction. Though it might be vulnerable to test ordering...?

cpennington commented:
Yeah, test ordering will change the query counts (and also, we don't have any evidence of how the cache affects the query counts, given this setup). It would be ideal to have a test for uncached, and a test for cached.

e0d commented:
@ormsbee what compression ratio were you seeing?  Our current max item size would by 1MB, so I'd like to know how much head-room we have and know what happens when/if that size is exceeded.

ormsbee commented:
I don't remember the ratio exactly, just that the largest course structure I could get my hands on based on the ticket compressed to about 400K. Right now, the failure behavior is bad -- my guess is that it will check to see if the record exists, find that it doesn't, attempt to set it, and then have memcached reject that. So we'll take the overhead of doing the compression work over and over again.

The compression ratios and how big our cached structure doc entries are would definitely be good to log and monitor via Datadog.

If we need to store compressed docs > 1MB, we could just have the caching code here do the chunking instead of doing config changes on the cluster. Though in the longer term, we need to move some stuff out of that structures doc. :(

adampalay commented:
@ormsbee , to @cpennington 's point about testing this can run the tests here without the cache (using django's DummyCache backend), then add a test explicitly _to_ test the cache?

adampalay commented:
@ormsbee , is this cache chunking an acceptance criterion?

ormsbee commented:
I would suggest it as follow-on work.

cpennington commented:
test this please

ormsbee commented on common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py:
```diff
@@ -256,15 +296,22 @@ def heartbeat(self):
+                with TIMER.timer("get_structure.find_one", course_context) as tagger_find_one:
+                    doc = self.structures.find_one({'_id': key})
+                    tagger_find_one.measure("blocks", len(doc['blocks']))
+                    structure = structure_from_mongo(doc)
```
@adampalay: why was `course_context` deleted from this call?

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py:
```diff
@@ -203,6 +209,40 @@ def structure_to_mongo(structure, course_context=None):
+        try:
+            self.cache = get_cache('course_structure_cache')
+        except InvalidCacheBackendError:
+            self.cache = get_cache('default')
```
I wasn't sure about using this or just not caching. This approach seemed like it could lead to unexpectedly bad cache behavior (if a site isn't ready for it), but I guess the other approach could lead to unexpectedly bad mongo behavior, so maybe it's a wash.

adampalay replied:
I worry about this, because we're not setting any expiration times on these really large documents. Will the regular cache quickly fill up and cause other things we want cached a little longer to be evicted?

Agreed it won't be an issue in our case, but I worry about other installations

ormsbee replied:
My understanding is that it's LRU by last access, so by definition whatever is getting evicted would be used less often than this. The biggest danger I can think of is this stomping on sessions for inactive users. It's a good argument for isolating the session caches from the content pieces, but I don't think changing the expiration time helps much.

adampalay commented:
failing test fixed in https://github.com/edx/edx-platform/pull/8605.

ormsbee commented:
Closing this in favor of #8632.
